yksgi/src/lib.rs

```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+
+                // Iterate over all locations. Register locations just update the current value in
+                // the registers vector. Direct locations make up part of the stack so need to
+                // written to the allocated memory. Other locations we haven't encountered yet, so
+                // will deal with them as they appear.
+                match l {
+                    SMLocation::Register(reg, _size) => {
+                        registers[usize::from(*reg)] = val;
+                    }
+                    SMLocation::Direct(reg, off, _) => {
+                        if i == 0 {
+                            // skip first frame
+                            continue;
+                        }
+                        debug_assert!(op.is_alloca());
+
+                        // The sizes reported by the stackmap aren't always correct. But we can get
+                        // the correct size from the IR.
+                        let eltype = unsafe { LLVMGetAllocatedType(op.get()) };
+                        let size = i64::try_from(unsafe { LLVMABISizeOfType(layout, eltype)}).unwrap();
+                        // Direct locations are always be in regards to RBP.
+                        assert_eq!(*reg, 6);
+                        let temp = unsafe { rbp.offset(isize::try_from(*off).unwrap()) };
+                        match size {
+                            4 => {
+                                let dval = unsafe { ptr::read(val as *mut u32) };
+                                unsafe { ptr::write(temp as *mut u32, dval as u32) };
+                            }
+                            8 => {
+                                let dval = unsafe { ptr::read(val as *mut u64) };
+                                unsafe { ptr::write(temp as *mut u64, dval) };
+                            }
+                            _ => { todo!() }
+                        }
+                    }
+                    SMLocation::Indirect(_reg, _off, _size) => {
+                        todo!()
+                    }
+                    SMLocation::Constant(_v) => {
+                        todo!()
+                    }
+                    SMLocation::LargeConstant(_v) => {
+                        todo!();
+                    }
+                }
+            }
+            if i > 0 {
+                // Advance the "virtual RSP" to the next frame.
+                rsp = unsafe { rbp.offset(-isize::try_from(rec.size).unwrap())};
+                if pinfo.hasfp {
+                    // The stack size recorded by the stackmap includes a pushed RBP. However, we
+                    // will have already adjusted the "virtual RSP" at the beginning if `hasfp` is
+                    // true. If that's the case, re-adjust the "virtual RSP" again to account for
+                    // this.
+                    rsp = unsafe { rsp.offset(8)};
+                }
+            }
+            // WRITE RETURN ADDRESS.
```

Comment:
Not sure we need this shouty line :)

Reply:
I did the caps on purpose to divide this code into sections and make them easier to find. Each caps title writes a different section to the new stack.